rskj-core/src/main/java/co/rsk/mine/MinerServerImpl.java

```diff
@@ -137,35 +151,211 @@ public MinerServerImpl(Ethereum ethereum,
 
         executor = new BlockExecutor(repository, blockchain, blockStore, null);
 
-        blocksWaitingforPoW = new LinkedHashMap<Sha3Hash, Block>(CACHE_SIZE) {
-            @Override
-            protected boolean removeEldestEntry(Map.Entry<Sha3Hash, Block> eldest) {
-                return size() > CACHE_SIZE;
-            }
-        };
+        blocksWaitingforPoW = createNewBlocksWaitingList();
 
         latestPaidFeesWithNotify = BigInteger.ZERO;
         latestParentHash = null;
         coinbaseAddress = miningConfig.getCoinbaseAddress();
         minFeesNotifyInDollars = BigDecimal.valueOf(miningConfig.getMinFeesNotifyInDollars());
         gasUnitInDollars = BigDecimal.valueOf(miningConfig.getMinFeesNotifyInDollars());
         minerMinGasPriceTarget = toBI(miningConfig.getMinGasPriceTarget());
+        timeBetweenBlocks = RskSystemProperties.CONFIG.getBlockchainConfig().getCommonConstants().getDurationLimit()*1000;
+        autoSwitchBetweenNormalAndPrivateMining = !RskSystemProperties.CONFIG.getBlockchainConfig().getCommonConstants().getPrivateMiningFallbackDifficulty().equals(BigInteger.ZERO);
+    }
+
+    private LinkedHashMap<Sha3Hash, Block> createNewBlocksWaitingList() {
+        return new LinkedHashMap<Sha3Hash, Block>(CACHE_SIZE)
+        {
+            @Override
+            protected boolean removeEldestEntry(Map.Entry<Sha3Hash, Block> eldest) {
+                return size() > CACHE_SIZE;
+            }
+        };
+
+    }
+
+    public int getPrivateBlocksGenerated() {
+     return privateBlocksGenerated;
+    }
+
+    public boolean isPrivateMining() {
+        return privateMining;
+    }
+
+    public void setPrivateMiningState() {
```

Comment:
Alternative implemention: move all the implementation to a new PrivateMinerClientImpl, so MinerServerImpl is agnostic of mining method. Maybe it could expose an alternative method to submitWork, to receive specially private mined blocks

Reply:
Yes, this is something to do in the future to improve the code.